### PR TITLE
Fix ability to upgrade chart

### PR DESCRIPTION
**Pull request by jbouse**
* Reworked NOTES.txt validation of passwords for upgrades
* Lint and improve the readability of the templates
* Update README to include upgrading

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -8,24 +8,44 @@
 
 ## About
 
-This is the Git repo of the official Helm charts for YOURLS.
+This is the Git repository of the official Helm charts for YOURLS.
```
Comment (LeoColomb) [inline README.md:11]: Not sure to understand this change.

File: README.md [hunk 2]
```diff
@@ -8,24 +8,44 @@
...
 
 ## Usage
 
-1. **Install Helm**  
+1. **Install Helm**
```
Comment (LeoColomb) [inline README.md:15]: Please don't remove trailing spaces on markdown.

File: README.md [hunk 3]
```diff
@@ -8,24 +8,44 @@
...
     Get the latest [Helm release](https://helm.sh/docs/intro/install/).
 
-2. **Add YOURLS Helm Chart repository**  
+2. **Add YOURLS Helm Chart repository**
```
Comment (LeoColomb) [inline README.md:18]: Please don't remove trailing spaces on markdown.

File: README.md [hunk 4]
```diff
@@ -8,24 +8,44 @@
...
+
     ```sh
     helm repo add yourls https://charts.yourls.org/
     helm repo update
     ```
 
-3. **Install a chart**  
+3. **Install a chart**
```
Comment (LeoColomb) [inline README.md:25]: Please don't remove trailing spaces on markdown.

File: README.md [hunk 5]
```diff
@@ -8,24 +8,44 @@
...
     * [Charts discovery](https://helm.sh/docs/helm/helm_search/)
-      ```sh
-      helm search yourls
-      ```
+
+    ```sh
+    helm search yourls
+    ```
+
+4. **Upgrade a chart**
```
Comment (LeoColomb) [inline README.md:35]: Why not, but this is specifically related to `yourls/yourls` chart and should live on `charts/yourls/README.md`.

File: charts/yourls/templates/_helpers.tpl [hunk 6]
```diff
@@ -45,65 +45,65 @@ Return the proper Docker Image Registry Secret Names
 Return the MySQL Hostname
 */}}
 {{- define "yourls.databaseHost" -}}
-{{- if .Values.mysql.enabled }}
-    {{- if eq .Values.mysql.architecture "replication" }}
-        {{- printf "%s-primary" (include "yourls.mysql.fullname" .) | trunc 63 | trimSuffix "-" -}}
+    {{- if .Values.mysql.enabled }}
```
Comment (LeoColomb) [inline charts/yourls/templates/_helpers.tpl:48]: By convention, the block opened by a `define` keeps the same indentation.
This is the same for the whole file.

File: charts/yourls/templates/NOTES.txt [hunk 7]
```diff
@@ -89,19 +89,15 @@ You can access Apache Prometheus metrics following the steps below:
 {{- include "common.warnings.rollingTag" .Values.image }}
 {{- include "common.warnings.rollingTag" .Values.metrics.image }}
 {{- include "common.warnings.rollingTag" .Values.volumePermissions.image }}
-{{- $passwordValidationErrors := list -}}
```
Comment (LeoColomb) [inline charts/yourls/templates/NOTES.txt:92]: The password validation array should stay outside of the actual validation for maintainability.

File: charts/yourls/templates/NOTES.txt [hunk 8]
```diff
@@ -89,19 +89,15 @@ You can access Apache Prometheus metrics following the steps below:
...
+
+  {{- if not (empty .Values.yourls.username) -}}
+    {{- $requiredPassword := dict "valueKey" "yourls.password" "secret" $secretName "field" "password" -}}
+    {{- $requiredPasswords = append $requiredPasswords $requiredPassword -}}
+  {{- end -}}
+
+  {{- $yourlsPasswordValidationErrors := include "common.validations.values.multiple.empty" (dict "required" $requiredPasswords "context" $) -}}
+  {{- include "common.errors.upgrade.passwords.empty" (dict "validationErrors" $yourlsPasswordValidationErrors "context" $) -}}
 {{- end }}
-{{- include "common.errors.upgrade.passwords.empty" (dict "validationErrors" $passwordValidationErrors "context" $) -}}
```
Comment (LeoColomb) [inline charts/yourls/templates/NOTES.txt:107]: The password validation function should stay outside of the actual validation for maintainability.

File: charts/yourls/templates/NOTES.txt [hunk 9]
```diff
@@ -89,19 +89,15 @@ You can access Apache Prometheus metrics following the steps below:
...
-{{- $passwordValidationErrors := list -}}
 {{- if not .Values.yourls.existingSecret -}}
-    {{- $secretName := include "yourls.secretName" . -}}
-    {{- $requiredYOURLSPassword := dict "valueKey" "yourls.password" "secret" $secretName "field" "password" "context" $ -}}
-    {{- $requiredYOURLSPasswordError := include "common.validations.values.single.empty" $requiredYOURLSPassword -}}
-    {{- $passwordValidationErrors = append $passwordValidationErrors $requiredYOURLSPasswordError -}}
-{{- end }}
-{{- if .Values.mysql.enabled }}
-    {{- $mysqlSecretName := include "yourls.databaseSecretName" . -}}
-    {{- $mysqlPasswordValidationErrors := "OK" -}}
```
Comment (LeoColomb) [inline charts/yourls/templates/NOTES.txt:101]: The issue is coming from here, the rest of the changes aren't needed.

**Review by LeoColomb — CHANGES_REQUESTED** [review]
Thanks for this second pull request @jbouse and for your investigation.
I'm only concerned about the many unrelated changes.

File: README.md (hunk 1 above)
Comment (jbouse) [reply README.md:11]: This was because of https://github.com/YOURLS/charts/runs/5355453994?check_suite_focus=true#step:4:135

Comment (LeoColomb) [reply README.md:11]: Oh ok, I see, thanks

File: README.md (hunk 2 above)
Comment (jbouse) [reply README.md:15]: This was editor syntax changes but I'll get them forced back.

**Comment by LeoColomb** [comment]
@jbouse I'm working on #123 right now, don't use too much of your time on this one. 😊

File: README.md (hunk 5 above)
Comment (jbouse) [reply README.md:35]: Oh I see what you mean... hadn't taken notice of the differences between the two README.md files in the repo

File: charts/yourls/templates/_helpers.tpl (hunk 6 above)
Comment (jbouse) [reply charts/yourls/templates/_helpers.tpl:48]: The `{{-` Jinja notation strips all left-hand white spacing. Just as `-}}` strips right-hand white spacing

File: charts/yourls/templates/NOTES.txt (hunk 9 above)
Comment (jbouse) [reply charts/yourls/templates/NOTES.txt:101]: Actually, I tested with and without each block and this block did not cause any issue with the validation.

**Comment by jbouse** [comment]
> @jbouse I'm working on #123 right now, don't use too much of your time on this one. 😊

I did see you trying to add `install --upgrade` to the `ct` call and I looked at that as well but the issue I ran into was being able to determine the secret name which meant having to set the passwords in the `*-values.yaml` file but that didn't seem to work either when upgrading.

**Pull request closed without merging**